Wearing colored glasses colors your vision, if you want to.

phil235 opened the pull request:
:cl: phil235
add: Wearing colored glasses colors your vision. This option can be enabled/disabled by alt-clicking any glasses with such feature.
/:cl:
- Wearing most colored glasses now tints your vision into a certain color if you have the preference enabled. It now uses /datum/client_colour and the client's color var.
- I added an altclick() option on all glasses to toggle a new pref regarding the use of glass coloring your view. The altclick toggles the preference, not the particular glasses you're wearing. Finally, this preference starts to off so you have to altclick some colored glasses (just once) to activate the effects on any colored glasses you might wear.
- I fixed items in our inventory UI being colored when using client.color. ( @RemieRichards )
- I fixed Thermal HUD Scanner not having its action button to switch hud type.

Health HUD glasses
![scrnsb](https://cloud.githubusercontent.com/assets/3476142/18594389/0ac9753a-7c40-11e6-8946-b335189d7e66.png)

Diagnostic HUD glasses
![yelllw](https://cloud.githubusercontent.com/assets/3476142/18594394/13451d9a-7c40-11e6-8f86-5a0dffcd0f9a.png)

sunglasses
![dark](https://cloud.githubusercontent.com/assets/3476142/18594398/152e626a-7c40-11e6-8c50-09716d6bbdd3.png)

Meson goggles
![lgreen](https://cloud.githubusercontent.com/assets/3476142/18594400/17d2d3fc-7c40-11e6-99cc-427c31e62419.png)

Science goggles
![purple](https://cloud.githubusercontent.com/assets/3476142/18594403/1955fdee-7c40-11e6-8779-e3e6c86fb6db.png)

Night Vision Goggles
![ggrre](https://cloud.githubusercontent.com/assets/3476142/18594407/1c185d4c-7c40-11e6-94f9-c06be5b6adeb.png)

phil235 committed:
Adds semi-transparent colored fullscreen overlays which are used by glasses with tint_color.
Currently only red, blue, and green are available. And only nightvision glasses currently have a tint_color when worn (green).

Dorsisdwarf commented:
Oooh.

GunHog commented:
I can see this getting annoying quite quickly.

Supermichael777 commented:
OW MY EYES

Ergovisavi commented:
If this doesn't have a similar utterly ridiculous "buh-DWEEEEEEEEET" sound from splinter cell I'll be sad.

TechnoAlchemisto commented:
this would be good if player perspective was different, seeing the big black overlays for places I can't see as well as this green tint kind of makes me nauseous

phil235 commented:
Would it be okay if I make it more transparent and less bright?

Supermichael777 commented:
use blend mode multiply instead of ad if you want to only color certain areas

PKPenguin321 commented:
Yes. Also put this on gas masks when

TheNightingale commented:
Maybe make it optional? That looks like the sort of thing to give me a headache real fast.

Dorsisdwarf commented:
>use blend mode multiply instead of ad if you want to only color certain areas
This. Bits you cant see should still be dark.

phil235 commented:
I added two other screenshots with less bright green tint.

Blend mode multiply doesn't work.

Bawhoppen commented:
needs to be more subtle, you know the red around your viewport that appears as you get damaged? needs to be like that but green

GunHog commented:
This could also be an idea:
![serveimage](https://cloud.githubusercontent.com/assets/4955510/18485119/1b28ac0c-79b1-11e6-8e31-8ff44565506e.jpg)

phil235 commented:
Well, if someone make a huge sprite for that, I can add it.

Niknakflak commented:
can you add a grey tint/greyscale to the detectives sunglasses somehow with this?

phil235 commented:
@Niknakflak Super easy to do, I only make NVGs use this new feature because people would complain less. But we can definitely add new tint color overlays to more glasses.

Durkel commented:
Could we possibly get a action button to toggle nvgs on/off if this gets merged?

ChangelingRain commented:
I feel like this should be color matrix based, not that I know how to do those

xxalpha commented:
You can't do it like that Joan.

Also this is a bad idea because the perspective is not first person.

RemieRichards commented:
@phil235 Why is this a huge screen object and not just a client colour datum?

Kromgar commented:
Ugh it looks terrible. I bet it looks even worse on lavaland.

MrPerson commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -531,6 +531,17 @@
+			overlay_fullscreen("color_tint", /obj/screen/fullscreen/color_vision/blue)
+		else
+			clear_fullscreen("color_tint", 0)
+
```
I feel this would be better if it just created the overlay and directly set its color to tintcolor rather than needing a new type of object whose only difference is one var

kilkun commented:
gas masks should also do this

MrPerson commented:
A simple tinting like this is very easy to do as a color matrix. I think what you want is something around
list(1,0,0,0, 0,1,0,0, 0,0,1,0, 0,0,0,1, 0,0.2,0,0)

Mathmatically 80 alpha w/ full green is 80/255=0.314~ but that looked far too green to me in quick testing. I liked the 0.2 a lot better. Plus people were bitching it was too bright.

pudl commented:
I dont like this unless you make it so that gas masks are green, science goggles are purple, etc. All colored glasses give tints. Otherwise its inconsistent and i dont like that

GunHog commented:
This should not obstruct vision or force to character to stare at a discolored screen for the entire duration they wear the goggles. 

How about a transition effect, similar to a flash, but muted and colored? For Science googles, you could have text scroll up on the screen, or for NVG, a green overlay that fades back to normal (along with a BWEEEE sound)!

Think of players like which constantly have a certain eyewear equipped as part of their role. If your screen had this the whole time as Nuke Ops, you would experience distortions as you re-adjust to the normal coloration after your round as an Op ends. Assuming you extend this feature to other eyewear, you could be forcing chemists and scientists to see purple shading on things for their entire rounds. Engineers and Miners seeing green constantly.

Please consider other methods of implementing this. I do not like the idea of having discolored memories of game objects because I wear colored glasses all round.

phil235 commented:
Okay, I'm testing client_colour datums, tell me which variation you like most (or none at all):

A
![7f7](https://cloud.githubusercontent.com/assets/3476142/18560518/3fbed0ba-7b7c-11e6-8324-da5ac386f277.png)

B
![5a5](https://cloud.githubusercontent.com/assets/3476142/18560526/435320dc-7b7c-11e6-8e14-270ce55dcb54.png)

C
![383](https://cloud.githubusercontent.com/assets/3476142/18560527/458fa820-7b7c-11e6-9157-5e4aadb60876.png)

D
![bfb](https://cloud.githubusercontent.com/assets/3476142/18560530/4957a638-7b7c-11e6-9b1a-9da773af4eb1.png)

E
![cfc](https://cloud.githubusercontent.com/assets/3476142/18560531/4a8c9518-7b7c-11e6-8a96-7d413624590f.png)

RemieRichards commented:
@phil235 I'm partial to the first one.

pudl commented:
@phil235 Option E is perfect. It makes it more based on aesthetics and doesnt make the game ugly

GunHog commented:
Agreed. "E" is the least detrimental option. I still feel there is a better way to do this, though. Seeing this is fine for a few minutes, but after that, it gets old.

Lexorion commented:
D and E are fine

Jalleo commented:
E is better in my opinion

MrStonedOne commented:
D

ChangelingRain commented:
D and E are nice, but I also like A somewhat.

TechnoAlchemisto commented:
I don't like that it tints everything but the UI, it's kind of weird to have the tinted items on top of the normally coloured UI.

MrStonedOne commented:
techno, thats what remie was talking about, we would make it remove the tint from the items inside the ui.

Dorsisdwarf commented:
E

Jalleo commented:
One thing I did suggest to Phil was perhaps making this a preference instead so everyone gets which one they want.

pudl commented:
Putting preferences on shit as minor as this is stupid, i think this is a good change

phil235 committed:
Completely rewrites how I did color tinting. We now use datum/client_colour and the client's color var.
I fixed items in our inventory UI being colored when using client.color.
I added an altclick() option on all glasses to toggle a new pref regarding the use of glass coloring your view.
I made a lot more glasses use vision coloring.
I fixed Thermal HUD Scanner not having its action button to switch hud type.

phil235 commented:
Alright, completely rewrote the feature, read the description and title.

nullbear commented:
I love the changes you've made to it.

pudl reviewed (changes requested):
dont make it toggleable

phil235 commented on code/_onclick/hud/fullscreen.dm:
```diff
@@ -106,3 +106,18 @@
+	color = "#ff0000"
+
+/obj/screen/fullscreen/color_vision/blue
+	color = "#0000ff"
```
forgot to remove this, since it's no longer used.

ChangelingRain commented:
I think the mesons, diag hud, and science goggles are a bit too visibile.

optimumtact commented:
It was a bad week in a bad month; maybe even a bad year. The Times was calling it the worst crime spike the city had ever seen. I’d just gone to Dino’s and ordered a quick bite, when in bursts Sergeant Sellecca—looking for me, as usual. Turns out Father McLeish got himself whacked. A gore-encrusted cricket bat found lying nearby matched the wounds on the departed. Job was done Thursday. I jotted down the details and got ready for a grisly new case.
     

It was raining; the kind of rain you’d hope would clean the filth right off the streets. Miss Bella Wood slinks through my door like a tigress sashays into a Burmese orphanage—a strawberry blonde with foot holders for hours and then some. No dame her age could afford a string of pearls like the one she wore, my money was right where my mind was: the gutter. She had bad news written on her like my last trip to Harlem. This dame was all business—before I could even close the office door, she lets slip that Ronald “the Rat” Arnold—he’d gone and gotten himself slain.
